versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java

```diff
@@ -387,24 +369,61 @@ protected Hash transplantAttempt(
             .collect(Collectors.toList());
 
     // 5. check for key-collisions
-    checkForKeyCollisions(ctx, targetHead, keysTouchedOnTarget, commitsToTransplantChronological);
+    return mergeTransplantCommon(
+        ctx,
+        timeInMicros,
+        targetHead,
+        branchCommits,
+        newKeyLists,
+        commitsToTransplantChronological,
+        targetEntriesReverseChronological,
+        transplantParams);
+  }
 
-    // (no need to verify the global states during a transplant)
-    // 6. re-apply commits in 'sequenceToTransplant' onto 'targetBranch'
-    targetHead =
-        copyCommits(
-            ctx,
-            timeInMicros,
-            targetHead,
-            commitsToTransplantChronological,
-            newKeyLists,
-            transplantParams.getUpdateCommitMetadata());
+  protected Hash mergeTransplantCommon(
+      OP_CONTEXT ctx,
+      long timeInMicros,
+      Hash toHead,
+      Consumer<Hash> branchCommits,
+      Consumer<Hash> newKeyLists,
+      List<CommitLogEntry> commitsToMergeChronological,
+      List<CommitLogEntry> toEntriesReverseChronological,
+      CopyCommitsParams params)
+      throws ReferenceConflictException, ReferenceNotFoundException {
 
-    // 7. Write commits
+    // Collect modified keys.
+    Collections.reverse(toEntriesReverseChronological);
+    Set<Key> keysTouchedOnTarget = collectModifiedKeys(toEntriesReverseChronological);
 
-    commitsToTransplantChronological.stream().map(CommitLogEntry::getHash).forEach(branchCommits);
-    writeMultipleCommits(ctx, commitsToTransplantChronological);
-    return targetHead;
+    checkForKeyCollisions(ctx, toHead, keysTouchedOnTarget, commitsToMergeChronological);
+
+    if (params.keepIndividualCommits() || commitsToMergeChronological.size() == 1) {
+      // (no need to verify the global states during a merge)
+      // re-apply commits in 'sequenceToTransplant' onto 'targetBranch'
+      toHead =
+          copyCommits(
+              ctx,
+              timeInMicros,
+              toHead,
+              commitsToMergeChronological,
+              newKeyLists,
+              m -> params.getUpdateCommitMetadata().apply(Collections.singletonList(m)));
```

Comment:
The relationship between the metadata rewrite function (with an implied `keepIndividualCommits` parameters ) and the explicit `keepIndividualCommits` parameters seems to be quite obscure to me.

Could we have a separate interface with _two_ methods for the metadata rewriter?

```
interface MetadataRewriter {
  CommitMeta rewriteSingle(CommitMeta metadata);
  CommitMeta squash(List<CommitMeta> metadata);
}
```

I suppose we'd only have to have one implementation this way - inside `BaseApiImpl`.

WDYT?

Reply:
Yea, looks much better that way.